prusti-specs/src/common.rs

```diff
@@ -1,29 +1,338 @@
+        fn rewrite_impl_item_method(&mut self, item: &mut ImplItemMethod) {
+            syn::visit_mut::visit_impl_item_method_mut(self, item);
+        }
+
+        fn rewrite_item_fn(&mut self, item: &mut ItemFn) {
+            syn::visit_mut::visit_item_fn_mut(self, item);
+        }
+
+        fn rewrite_tokens(&self, tokens: TokenStream) -> TokenStream {
+            let tokens_span = tokens.span();
+            let rewritten = TokenStream::from_iter(tokens.into_iter().map(|token| match token {
+                TokenTree::Group(group) => {
+                    let new_group =
+                        proc_macro2::Group::new(group.delimiter(), self.rewrite_tokens(group.stream()));
+                    TokenTree::Group(new_group)
+                }
+                TokenTree::Ident(ident) if ident == "self" => {
+                    TokenTree::Ident(proc_macro2::Ident::new("_self", ident.span()))
+                },
+                other => other,
+            }));
+            parse_quote_spanned! {tokens_span=>
+            #rewritten
+        }
+        }
+    }
+
+    impl<'a, T: ToTokens> VisitMut for Rewriter<'a, T> {
+        fn visit_fn_arg_mut(&mut self, fn_arg: &mut FnArg) {
+            if let FnArg::Receiver(receiver) = fn_arg {
+                let span = receiver.span();
+                let and = if receiver.reference.is_some() {
+                    // TODO: do lifetimes need to be specified here?
+                    quote_spanned! {span=> &}
+                } else {
+                    quote! {}
+                };
+                let mutability = &receiver.mutability;
+                let new_ty = self.new_ty;
+                let new_fn_arg: FnArg = parse_quote_spanned! {span=>
+                _self : #and #mutability #new_ty
+            };
+                *fn_arg = new_fn_arg;
+            } else {
+                syn::visit_mut::visit_fn_arg_mut(self, fn_arg);
+            }
+        }
+
+        fn visit_ident_mut(&mut self, ident: &mut Ident) {
+            if ident == "self" {
+                *ident = Ident::new("_self", ident.span());
+            }
+            syn::visit_mut::visit_ident_mut(self, ident)
+        }
+
+        fn visit_macro_mut(&mut self, makro: &mut Macro) {
+            // A macro can appear in a spec function (e.g. `matches!`)
+            makro.tokens = self.rewrite_tokens(makro.tokens.clone());
+            syn::visit_mut::visit_macro_mut(self, makro);
+        }
     }
 }
 
-impl HasGenerics for syn::ItemImpl {
-    fn get_generics(&self) -> &syn::Generics {
-        &self.generics
+/// Copies the [syn::Generics] of `source` to the generics of `target`
+pub(crate) fn merge_generics<T: HasGenerics>(target: &mut T, source: &T) {
+    let generics_target = target.generics_mut();
+    let generics_source = source.generics();
+
+    generics_target.params.extend(generics_source.params.clone());
+    if let (Some(target_where), Some(source_where)) =
```

Comment:
What if `generics_target.where_clause` is `None` but `generics_source.where_clause` is `Some(_)`? Then `generics_target` should become `Some(_)`, right?

Reply:
Indeed! Also, type params could be duplicated. I fixed this and added some tests for it.